sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CoderRegistry.java

```diff
@@ -82,34 +89,67 @@
+    codersToRegister.put(ByteString.class, CoderFactories.fromStaticMethods(ByteStringCoder.class));
+    codersToRegister.put(Double.class, CoderFactories.fromStaticMethods(DoubleCoder.class));
+    codersToRegister.put(Instant.class, CoderFactories.fromStaticMethods(InstantCoder.class));
+    codersToRegister.put(Integer.class, CoderFactories.fromStaticMethods(VarIntCoder.class));
+    codersToRegister.put(Iterable.class, CoderFactories.fromStaticMethods(IterableCoder.class));
+    codersToRegister.put(KV.class, CoderFactories.fromStaticMethods(KvCoder.class));
+    codersToRegister.put(List.class, CoderFactories.fromStaticMethods(ListCoder.class));
+    codersToRegister.put(Long.class, CoderFactories.fromStaticMethods(VarLongCoder.class));
+    codersToRegister.put(Map.class, CoderFactories.fromStaticMethods(MapCoder.class));
+    codersToRegister.put(Set.class, CoderFactories.fromStaticMethods(SetCoder.class));
+    codersToRegister.put(String.class, CoderFactories.fromStaticMethods(StringUtf8Coder.class));
+    codersToRegister.put(TimestampedValue.class,
+        CoderFactories.fromStaticMethods(TimestampedValue.TimestampedValueCoder.class));
+    codersToRegister.put(Void.class, CoderFactories.fromStaticMethods(VoidCoder.class));
+    codersToRegister.put(byte[].class, CoderFactories.fromStaticMethods(ByteArrayCoder.class));
+    codersToRegister.put(IntervalWindow.class, CoderFactories.forCoder(IntervalWindow.getCoder()));
+
+    for (CoderRegistrar registrar : registrars) {
+      for (Map.Entry<Class<?>, CoderFactory> entry
+          : registrar.getCoderFactoriesToUseForClasses().entrySet()) {
+        codersToRegister.put(entry.getKey(), entry.getValue());
+      }
+    }
+
+    // Warn the user if multiple coders have been registered for the same class
+    Map<Class<?>, Collection<CoderFactory>> multipleRegistrations =
+        Maps.filterValues(codersToRegister.asMap(), new Predicate<Collection<CoderFactory>>() {
+      @Override
+      public boolean apply(@Nonnull Collection<CoderFactory> input) {
+        return input.size() > 1;
+      }
+    });
+    for (Map.Entry<Class<?>, Collection<CoderFactory>> entry : multipleRegistrations.entrySet()) {
+      LOG.warn("Multiple CoderFactory registrations {} found for class {}, using {}.",
+          entry.getKey(), entry.getValue(), entry.getValue().iterator().next());
+    }
+
+    ImmutableMap.Builder<Class<?>, CoderFactory> registeredCoderFactoriesPerClassBuilder =
+        ImmutableMap.builder();
+    for (Map.Entry<Class<?>, Collection<CoderFactory>> entry
+        : codersToRegister.asMap().entrySet()) {
+      registeredCoderFactoriesPerClassBuilder.put(
+          entry.getKey(), entry.getValue().iterator().next());
+    }
+    REGISTERED_CODER_FACTORIES_PER_CLASS = registeredCoderFactoriesPerClassBuilder.build();
+  }
 
   public CoderRegistry() {
+    coderFactoryMap = new HashMap<>(REGISTERED_CODER_FACTORIES_PER_CLASS);
```

Comment:
A pipeline author will never create a registry themselves at all. And after your change, they won't even interact with it directly, and probably IO/transform authors will not either.

So all of these are aimed at runner authors, testing (hermeticity), and just good compositional modeling so that it can be used for things we may not have thought of yet.

They can always be added later, so just choosing a meaningful factory method name so there's room for others is enough. I am OK with `createDefault`, but the characters in `create` don't add any value so maybe there's a better synonym to avoid the reserved word `default`.